File: CPP/petshop.cpp
#include <iostream>
#include <string>
#include <list>     // Pake list biar bisa pushback

using namespace std;

class petshop {
private:
    int id;
    string nama;
    string kategori;
    int harga;
    int stok;

public:
    petshop() {
        this->id = 0;
        this->nama = "";
        this->kategori = "";
        this->harga = 0;
        this->stok = 0;
    }

    petshop(int id, string nama, string kategori, int harga, int stok) {
        this->id = id;
        this->nama = nama;
        this->kategori = kategori;
        this->harga = harga;
        this->stok = stok;
    }

    int get_id() { return id; }
    void set_id(int id) { this->id = id; }
    string get_nama() { return nama; }
    void set_nama(string nama) { this->nama = nama; }
    string get_kategori() { return kategori; }
    void set_kategori(string kategori) { this->kategori = kategori; }
    int get_harga() { return harga; }
    void set_harga(int harga) { this->harga = harga; }
    int get_stok() { return stok; }
    void set_stok(int stok) { this->stok = stok; }

    class Menu {
    private:
        list<petshop> data;

    public:
        Menu() {}

        void tampilkanInfo(petshop pet) {
            cout << "ID: " << pet.get_id() << endl;
            cout << "Nama Produk: " << pet.get_nama() << endl;
            cout << "Kategori: " << pet.get_kategori() << endl;
            cout << "Harga: " << pet.get_harga() << endl;
            cout << "Stok: " << pet.get_stok() << endl;
            cout << "-------------------------" << endl;
        }

        void Create() {
            int id, harga, stok;
            string nama, kategori;

            cout << "Masukkan ID: ";
            cin >> id;
            cout << "Masukkan Nama Produk: ";
            cin >> nama;
            cout << "Masukkan Kategori: ";
            cin >> kategori;
            cout << "Masukkan Harga: ";
            cin >> harga;
            cout << "Masukkan Stok: ";
            cin >> stok;

            data.push_back(petshop(id, nama, kategori, harga, stok));
            cout << "Data berhasil ditambahkan!" << endl;
        }

        void Print() {
            if (data.empty()) {
                cout << "Tidak ada data yang tersedia." << endl;
            } else {
                cout << "-------------------------" << endl;
                cout << "Berikut Peralatan PetShop" << endl;
                cout << "-------------------------" << endl;
                for (list<petshop>::iterator it = data.begin(); it != data.end(); ++it) {
                    tampilkanInfo(*it);
                }
            }
        }

        void Update() {
            int id;
            cout << "Masukkan ID produk yang ingin diubah: ";
            cin >> id;

            bool ditemukan = false;
            for (list<petshop>::iterator it = data.begin(); it != data.end(); ++it) {
                if (it->get_id() == id) {
                    string nama, kategori;
                    int harga, stok;

                    cout << "Masukkan Nama Produk Baru: ";
                    cin >> nama;
                    cout << "Masukkan Kategori Baru: ";
                    cin >> kategori;
                    cout << "Masukkan Harga Baru: ";
                    cin >> harga;
                    cout << "Masukkan Stok Baru: ";
                    cin >> stok;

                    it->set_nama(nama);
                    it->set_kategori(kategori);
                    it->set_harga(harga);
                    it->set_stok(stok);

                    cout << "Data berhasil diubah!" << endl;
                    ditemukan = true;
                    break;
                }
            }

            if (!ditemukan) {
                cout << "Data dengan ID " << id << " tidak ditemukan." << endl;
            }
        }

        void Delete() {
            int id;
            cout << "Masukkan ID produk yang ingin dihapus: ";
            cin >> id;

            list<petshop>::iterator it = data.begin();
            while (it != data.end()) {
                if (it->get_id() == id) {
                    it = data.erase(it);
                    cout << "Data berhasil dihapus!" << endl;
                    return;
                }
                ++it;
            }
            cout << "Data dengan ID " << id << " tidak ditemukan." << endl;
        }

        void Search() {
            string nama;
            cout << "Masukkan Nama Produk yang ingin dicari: ";
            cin >> nama;

            bool ditemukan = false;
            for (list<petshop>::iterator it = data.begin(); it != data.end(); ++it) {
                if (it->get_nama() == nama) {
                    tampilkanInfo(*it);
                    ditemukan = true;
                }
            }

            if (!ditemukan) {
                cout << "Data dengan nama " << nama << " tidak ditemukan." << endl;
            }
        }
        ~Menu() {}
    };
    ~petshop() {}
};

File: Python/Petshop.py
class Petshop:
    def __init__(self, id=0, nama="", kategori="", harga=0, stok=0):
        # Constructor untuk kelas Petshop,
        # Sekaligus menginisialisasi atribut dasar seperti id, nama, kategori, harga, dan stok
        self.id = id
        self.nama = nama
        self.kategori = kategori
        self.harga = harga
        self.stok = stok

    # Getter dan setter untuk setiap atribut
    def get_id(self):
        return self.id

    def set_id(self, id):
        self.id = id

    def get_nama(self):
        return self.nama

    def set_nama(self, nama):
        self.nama = nama

    def get_kategori(self):
        return self.kategori

    def set_kategori(self, kategori):
        self.kategori = kategori

    def get_harga(self):
        return self.harga

    def set_harga(self, harga):
        self.harga = harga

    def get_stok(self):
        return self.stok

    def set_stok(self, stok):
        self.stok = stok

    # Kelas Menu untuk mengelola operasi CRUD (Create, Read, Update, Delete) pada data Petshop
    class Menu:
        def __init__(self):
            # Inisialisasi list kosong untuk menyimpan data produk
            self.data = []

        def tampilkan_info(self, pet):
            # Menampilkan informasi detail dari sebuah produk
            print(f"ID: {pet.get_id()}")
            print(f"Nama Produk: {pet.get_nama()}")
            print(f"Kategori: {pet.get_kategori()}")
            print(f"Harga: {pet.get_harga()}")
            print(f"Stok: {pet.get_stok()}")
            print("-------------------------")

        def create(self):
            # Method untuk menambahkan data produk baru ke dalam list
            id = int(input("Masukkan ID: "))
            nama = input("Masukkan Nama Produk: ")
            kategori = input("Masukkan Kategori: ")
            harga = int(input("Masukkan Harga: "))
            stok = int(input("Masukkan Stok: "))

            # Membuat objek Petshop baru dan menambahkannya ke list data
            self.data.append(Petshop(id, nama, kategori, harga, stok))
            print("Data berhasil ditambahkan!")

        def print(self):
            # Method untuk menampilkan semua data produk yang tersimpan
            if not self.data:
                print("Tidak ada data yang tersedia.")
            else:
                print("-------------------------")
                print("Berikut Peralatan PetShop")
                print("-------------------------")
                for pet in self.data:
                    self.tampilkan_info(pet)

        def update(self):
            # Method untuk mengubah data produk berdasarkan ID
            id = int(input("Masukkan ID produk yang ingin diubah: "))
            ditemukan = False

            for pet in self.data:
                if pet.get_id() == id:
                    # Jika ID ditemukan, minta input baru untuk setiap atribut
                    nama = input("Masukkan Nama Produk Baru: ")
                    kategori = input("Masukkan Kategori Baru: ")
                    harga = int(input("Masukkan Harga Baru: "))
                    stok = int(input("Masukkan Stok Baru: "))

                    # Update nilai atribut menggunakan setter
                    pet.set_nama(nama)
                    pet.set_kategori(kategori)
                    pet.set_harga(harga)
                    pet.set_stok(stok)

                    print("Data berhasil diubah!")
                    ditemukan = True
                    break

            if not ditemukan:
                print(f"Data dengan ID {id} tidak ditemukan.")

        def delete(self):
            # Method untuk menghapus data produk berdasarkan ID
            id = int(input("Masukkan ID produk yang ingin dihapus: "))
            for pet in self.data:
                if pet.get_id() == id:
                    self.data.remove(pet)
                    print("Data berhasil dihapus!")
                    return

            print(f"Data dengan ID {id} tidak ditemukan.")

        def search(self):
            # Method untuk mencari data produk berdasarkan nama
            nama = input("Masukkan Nama Produk yang ingin dicari: ")
            ditemukan = False

            for pet in self.data:
                if pet.get_nama() == nama:
                    self.tampilkan_info(pet)
                    ditemukan = True

            if not ditemukan:
                print(f"Data dengan nama {nama} tidak ditemukan.")